### ISSUE TITLE
csi-driver/provisioner for kubernetes pv/pvc  creation. 

### ISSUE THREAD

**Issue by yydzhou** [Alluxio/alluxio#9599]
**Is your feature request related to a problem? Please describe.**
Current Alluxio supports basic usage in kubernetes environment like s3 api or POSIX interface via FUSE. But there is no provisioner/csi-driver that can create pv/pvc of alluxio and make storage transparent to Kubernetes. The gap make current kuberentes users who are using pv/pvc hard to integrate Alluxio into their current deployment.
 
CSI spec link: https://github.com/container-storage-interface/spec
Kubernetes CSI support status: https://kubernetes.io/blog/2019/01/15/container-storage-interface-ga/
A CSI-driver based provisioner provides the following benefits. 
- Match kubernetes design as PV/PVC is kubernetes standard storage solution/policy for data access/management. So current kubernetes users can integrate alluxio into their deployment with little/no effort. 
- PV/PVC already has the data access/security control. So user don't need to implement/find another one.
- Loose the host dependency, like hostpath configuration, library installation. 
- CSI is the standard of data/storage interface in cloud environment. So one implementation can enable the Alluxio be used in other Orchestration Systems. 
   
**Describe the solution you'd like**
Implement a csi driver/provisioner that can create/manage/mount persistent volume (PV) in kubernetes. So user can just deploy the provisioner and a storageclass to consume/use the alluxio.
Here is an example of csi: https://github.com/ceph/ceph-csi

**Describe alternatives you've considered**
Make/push kubernetes to support Alluxio persistent volume like https://kubernetes.io/docs/concepts/storage/persistent-volumes/#types-of-persistent-volumes
So that user can configure persistent volume manually/statically and switch their current deployment to Alluxio based storage. 

**Urgency**
Explain why the feature is important

**Additional context**

**Comment by mingfang** [comment]
I created a CSI for Alluxio https://github.com/mingfang/alluxio-csi

**Comment by madanadit** [comment]
@mingfang Would you like to migrate the csi plugin to the alluxio repository and add documentation? we can create a dir called `csi` under https://github.com/Alluxio/alluxio/tree/master/integration/kubernetes. I'll be more than happy to help push the PR. Thanks for your contribution!

**Comment by mingfang** [comment]
@madanadit Yes let's do it.

**Comment by madanadit** [comment]
@mingfang Great! Would you like to create the PR and I can help get it in? Let me know what works for you.

**Comment by mingfang** [comment]
@madanadit The CSI is written in Golang. It's better for it to be on its own repo, e.g. Alluxio/alluxio-csi

**Comment by madanadit** [comment]
Thanks for the suggestion @mingfang. Let me circulate this idea with our go experts and revert

**Comment by Xenorith** [comment]
Hi @mingfang 

I took a look at your CSI code and I agree that we should create a separate repo to host the code. I'd like to make a few suggestions to make the code more accessible and minimize the number of dependencies:
- We currently use `go modules` to manage dependencies; this tool is available starting go 1.11. Is it possible for you to migrate the dependency management from `go dep`? For reference, see https://github.com/golang/go/wiki/Modules#migrating-to-modules
- The Gopkg.toml file only specifies 3 dependencies (yay!). I assume `kubernetes` and `container-storage-interface` are both critical dependencies, but is it possible to avoid using `glog` in favor of go's native logging library?
- Similarly, I noticed the use of grpc's `status` and `codes` packages for error propagation. Is it possible to remove the import and use native error libraries?

Thanks again for all your work to get this started!

**Comment by madanadit** [comment]
Hi guys, here is the new repo: https://github.com/Alluxio/alluxio-csi. @mingfang do you mind opening a PR to push your code? Thanks again for your contribution!

**Comment by mingfang** [comment]
Yes, I will make the changes suggested by @Xenorith and then open a PR.

**Comment by madanadit** [comment]
@mingfang do you need any discussion / input from us to make progress? thx
